src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -203,6 +214,93 @@ public PostingsReader getPostingReader(long offset) throws IOException
+        private final QueryContext context;
+
+        private final Expression exp;
+
+        RangeQuery(Expression exp, QueryEventListener.TrieIndexEventListener listener, QueryContext context)
+        {
+            this.listener = listener;
+            this.exp = exp;
+            lookupStartTime = System.nanoTime();
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            final ByteComparable lower = exp.lower != null ? ByteComparable.fixedLength(exp.getLowerBound()) : null;
+            final ByteComparable upper = exp.upper != null ? ByteComparable.fixedLength(exp.getUpperBound()) : null;
+            try (TrieRangeIterator reader = new TrieRangeIterator(termDictionaryFile.instantiateRebufferer(),
+                                                                  termDictionaryRoot,
+                                                                  lower,
+                                                                  upper,
+                                                                  exp.lowerInclusive,
+                                                                  exp.upperInclusive))
+            {
+                var iter = reader.iterator();
+                if (!iter.hasNext())
+                    return PostingList.EMPTY;
+
+                context.checkpoint();
+                // Because postings are not sorted, we need to eagerly materialize the results and sort them.
+                LongHeap postings = materializeResults(iter);
+
+                listener.onTraversalComplete(System.nanoTime() - lookupStartTime, TimeUnit.NANOSECONDS);
+
+                return new LongHeapPostingList(postings);
+            }
+            catch (Throwable e)
+            {
+                if (!(e instanceof AbortedOperationException))
+                    logger.error(indexContext.logMessage("Failed to execute term query"), e);
+
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        /**
+         * Build an in-memory heap of row ids from the posting lists of the matching terms.
+         * @return an ordered {@link LongHeap} of row ids
+         */
+        private LongHeap materializeResults(Iterator<Pair<ByteSource,Long>> trieEntries) throws IOException
+        {
+            assert trieEntries.hasNext();
+            var heap = new LongHeap(1);
+            try (var postingsInput = IndexFileUtils.instance.openInput(postingsFile);
+                 var postingsSummaryInput = IndexFileUtils.instance.openInput(postingsFile))
+            {
+                do
+                {
+                    Pair<ByteSource, Long> nextTrieEntry = trieEntries.next();
+                    ByteSource key = nextTrieEntry.left;
+                    long rowId = nextTrieEntry.right;
+                    byte[] nextBytes = ByteSourceInverse.readBytes(key);
+                    if (exp.isSatisfiedBy(ByteBuffer.wrap(nextBytes)))
```

Comment:
Under what circumstances would this return `false`? If I comment this condition out, all the tests still pass. 

A follow on to this is, if it can return `false`, is it valid to just ignore the entry?

Reply:
> Under what circumstances would this return `false`? If I comment this condition out, all the tests still pass.

The current logic in the class is to get all of the map entries with a given prefix, and then we filter them here. We "have" to do this because the trie is not encoded in the correct order. When the map entry's value is encoded, it is always as a literal, which does not produce the right results for numbers unless we get all of the entries for a given key. We filter here to prevent sending unnecessary rows to be read from disk. 

The tests pass because this is an optimization, not a logical requirement. When this line is commented out, the row filter will filter out more rows.

The correct long term plan is to fix the way the trie is encoded for map entries so that we can push the selection logic down to the `TrieRangeIterator`.

> A follow on to this is, if it can return `false`, is it valid to just ignore the entry?

We ignore the entry because it doesn't satisfy the query predicate.

> If this test isn't needed then we don't need to have the iterator return a `Pair<ByteSource, Long>`. It could just return a `Long`. We could then simplify the `TrieRangeIterator` to not have the `collector`. This would avoid a lot of `ByteSource` object creation.

Yes, this is my plan once we fix the trie encoding. I think this comes down to whether the feature needs to be backwards compatible. The goal was to get something working quickly, so this solution uses more memory in order to be ready sooner. My understanding from @jbellis is that we can optimize this based on feature usage.

Reply:
> We could then simplify the `TrieRangeIterator` to not have the `collector`

For what it's worth, this might be a hidden problem. The collector appears to build a stack of the results in memory, which could be another place for memory pressure. I haven't looked into the design much, but it is yet another reason why a serious usage of this feature would greatly benefit from fixing the trie encoding for map entries.

Reply:
Ok, I understand now. Thanks. I think I was confused by the naming of the `ByteSource` as `key`. In my mind this translated to the map key. This then turns into a nit that this may be better named as `entry` or `entryBytes`.

Reply:
Good point. Those names are not very good, I'll update them.

Reply:
> For what it's worth, this might be a hidden problem. 

This is solved in 0a9a861ac03e663f3a6c350d57b0797f33e2a8a2. We can now use `TrieTermsDictionaryReader`, which was just updated to implement `ValueIterator`, which lazily materializes ByteComparable objects.